pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,667 @@
+				fmt.Fprintf(o.ErrOut, "info: Filtered all images from %s, skipping\n", src.ref)
+				continue
+			}
+
+			for _, dst := range pushTargets {
+				// if we are going to be using cross repository mount, get a token that covers the src
+				if src.ref.Registry == dst.ref.Registry {
+					toContext = toContext.WithScopes(auth.RepositoryScope{Repository: src.ref.RepositoryName(), Actions: []string{"pull"}})
+				}
+
+				toRepo, err := o.Repository(ctx, toContext, creds, dst.t, dst.ref)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to connect to %s: %v", dst.ref, err)})
+					continue
+				}
+
+				canonicalTo := toRepo.Named()
+				toManifests, err := toRepo.Manifests(ctx)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to access destination image %s manifests: %v", src.ref, err)})
+					continue
+				}
+
+				// if the destination tag already has this manifest, do nothing
+				var mustCopyLayers bool
+				if o.Force {
+					mustCopyLayers = true
+				} else {
+					if _, err := toManifests.Get(ctx, srcDigest); err != nil {
+						mustCopyLayers = true
+					} else {
+						glog.V(4).Infof("Manifest exists in %s, no need to copy layers without --force", dst.ref)
+					}
+				}
+				if mustCopyLayers {
+					// upload all the blobs
+					toBlobs := toRepo.Blobs(ctx)
+					srcBlobs := srcRepo.Blobs(ctx)
+
+					// upload the each manifest
+					for _, srcManifest := range srcManifests {
+						switch srcManifest.(type) {
+						case *schema2.DeserializedManifest:
+						default:
+							digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("the manifest type %T is not supported", srcManifest)})
+							continue
+						}
+
+						for _, blob := range srcManifest.References() {
+							// tagging within the same registry is a no-op
+							if src.ref.Registry == dst.ref.Registry && canonicalFrom.String() == canonicalTo.String() {
+								continue
+							}
+
+							blobSource, err := reference.WithDigest(canonicalFrom, blob.Digest)
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unexpected error building named digest: %v", err)})
+								continue
+							}
+
+							// if we aren't forcing upload, skip the blob copy
+							if !o.Force {
+								_, err := toBlobs.Stat(ctx, blob.Digest)
+								if err == nil {
+									// blob exists, skip
+									glog.V(5).Infof("Server reports blob exists %#v", blob)
+									continue
+								}
+								if err != distribution.ErrBlobUnknown {
+									glog.V(5).Infof("Server was unable to check whether blob exists %s: %v", blob.Digest, err)
+								}
+							}
+
+							var options []distribution.BlobCreateOption
+							if !o.SkipMount {
+								options = append(options, client.WithMountFrom(blobSource), WithDescriptor(blob))
```

Comment:
This MountFrom makes sense only when src and dst registries are the same.

As an optimization, pushed/skipped blobs to the same registry could be remembered and then used for MountFrom.

Reply:
Registries must ignore this field when it's not valid, so it doesn't cost us anything to always send it.

Reply:
Please put there at least a comment that this isn't supposed to work in most cases.